ext/drx/drx_time_scale.c

```diff
@@ -500,47 +512,64 @@ scale_posix_timers(void *drcontext, bool inflate)
-                    deflate_timespec(drcontext, &spec.it_interval);
-                    deflate_timespec(drcontext, &spec.it_value);
-                }
-                int res = dr_invoke_syscall_as_app(drcontext, SYS_timer_settime, 4, id, 0,
-                                                   &spec, NULL);
-                if (res != 0)
-                    NOTIFY(0, "Failed to call timer_settime for id %d: %d\n", id, res);
-            } else if (res != 0) {
+            res = dr_invoke_syscall_as_app(drcontext, SYS_timer_gettime, 2, id, &spec);
+            if (res != 0) {
                 NOTIFY(0, "Failed to call timer_gettime for id %d: %d\n", id, res);
+                continue;
+            }
+            if (is_timespec_zero(&spec.it_value) && is_timespec_zero(&spec.it_interval)) {
+                /* Disabled; nothing to do. */
+                continue;
+            }
+            /* If the timer just fired, we can have a zero it_value.  If we send that
+             * to timer_settime we will disable the timer, so we reset it ourselves.
+             */
+            if (is_timespec_zero(&spec.it_value)) {
+                spec.it_value.tv_sec = spec.it_interval.tv_sec;
+                spec.it_value.tv_nsec = spec.it_interval.tv_nsec;
             }
+            if (inflate) {
+                inflate_timespec(drcontext, &spec.it_interval);
+                inflate_timespec(drcontext, &spec.it_value);
+            } else {
+                deflate_timespec(drcontext, &spec.it_interval);
+                deflate_timespec(drcontext, &spec.it_value);
+            }
+            res = dr_invoke_syscall_as_app(drcontext, SYS_timer_settime, 4, id, 0, &spec,
+                                           NULL);
+            if (res != 0)
+                NOTIFY(0, "Failed to call timer_settime for id %d: %d\n", id, res);
 #else
             itimerspec64_t spec;
             /* See above comment about dr_invoke_syscall_as_app(). */
-            int res =
-                dr_invoke_syscall_as_app(drcontext, SYS_timer_gettime64, 2, id, &spec);
-            if (res == 0 &&
-                (!is_timespec64_zero(&spec.it_interval) ||
-                 !is_timespec64_zero(&spec.it_value))) {
-                if (inflate) {
-                    inflate_timespec64(drcontext, &spec.it_interval);
-                    inflate_timespec64(drcontext, &spec.it_value);
-                } else {
-                    deflate_timespec64(drcontext, &spec.it_interval);
-                    deflate_timespec64(drcontext, &spec.it_value);
-                }
-                int res = dr_invoke_syscall_as_app(drcontext, SYS_timer_settime64, 4, id,
-                                                   0, &spec, NULL);
-                if (res != 0)
-                    NOTIFY(0, "Failed to call timer_settime64 for id %d: %d\n", id, res);
-            } else if (res != 0) {
+            res = dr_invoke_syscall_as_app(drcontext, SYS_timer_gettime64, 2, id, &spec);
```

Comment:
This is a bit confusing: `itimerspec64_t` and `SYS_timer_[get | set]time64` in the #else (i.e., in the 32 bit section)?
Looks like that's what we want though (looking at: https://github.com/DynamoRIO/dynamorio/blob/master/ext/drx/drx_time_scale.c#L248).
Just makes me wonder if this is the right naming convention (having 64 in the name)?